kyo-core/shared/src/main/scala/kyo/StreamCoreExtensions.scala

```diff
@@ -533,6 +531,82 @@ object StreamCoreExtensions:
         ): Stream[V2, Abort[E] & Async & S & S2] =
             mapChunkParUnordered(Async.defaultConcurrency, defaultAsyncStreamBufferSize)(f)(using t1, t2, t3, i1, i2, ev, frame)
 
+        /** Collects values that are emitted within the same duration [[maxTime]] or reach the amount [[maxSize]] and emits them as a chunk.
+          *
+          * @param maxSize
+          *   Maximum number of elements to be collected within a single duration
+          * @param maxTime
+          *   Maximum amount of time to collect elements to emit in chunk
+          * @return
+          *   A new stream that emits collected chunks of elements
+          */
+        def groupedWithin(maxSize: Int, maxTime: Duration)(using
+            t1: Tag[Emit[Chunk[V]]],
+            t2: Tag[V],
+            i1: Isolate.Contextual[S, IO],
+            i2: Isolate.Stateful[S, Abort[E] & Async],
+            fr: Frame
+        ): Stream[Chunk[V], S & Async] =
+            Stream:
+                AtomicRef.initWith(Chunk.empty[V]): bufferRef =>
+                    Channel.initWith[Maybe[Chunk[V]]](defaultAsyncStreamBufferSize): channel =>
+                        IO.ensure(channel.close):
+                            AtomicRef.initWith(Fiber.unit[Closed]): fiberRef =>
+                                IO.ensure(fiberRef.get.map(_.interrupt)):
+                                    val flushLoop: Unit < (Async & Abort[Closed]) =
+                                        Loop.foreach:
+                                            Async.sleep(maxTime).andThen:
```

Comment:
It seems the flushing can drift depending on the timing of the arrival of new items and if there's already a flush loop running or not. Maybe `Clock.repeat*` could be more predictable?